enterprise/server/raft/replica/replica.go

```diff
@@ -1208,31 +1212,31 @@ func (sm *Replica) storedFileWriter(ctx context.Context, fileRecord *rfpb.FileRe
 }
 
 func (sm *Replica) Writer(ctx context.Context, header *rfpb.Header, fileRecord *rfpb.FileRecord) (interfaces.CommittedWriteCloser, error) {
-	db, err := sm.leaser.DB()
-	if err != nil {
-		return nil, err
-	}
-
 	if err := sm.validateRange(header); err != nil {
-		db.Close()
 		return nil, err
 	}
 
 	fileMetadataKey, err := sm.fileStorer.FileMetadataKey(fileRecord)
 	if err != nil {
-		db.Close()
 		return nil, err
 	}
 
 	writeCloserMetadata, err := sm.storedFileWriter(ctx, fileRecord)
 	if err != nil {
-		db.Close()
 		return nil, err
 	}
 
 	wc := ioutil.NewCustomCommitWriteCloser(writeCloserMetadata)
-	wc.CloseFn = db.Close
 	wc.CommitFn = func(bytesWritten int64) error {
+		db, err := sm.leaser.DB()
+		if err != nil {
+			if sm.IsSplitting() {
+				return status.OutOfRangeErrorf("%s: id %d ", constants.RangeSplittingMsg, header.GetRangeId())
+			}
+			return err
+		}
+		defer db.Close()
+
 		batch := db.NewBatch()
```

Comment:
do we need to revalidate the header? it's possible the range descriptor has changed between the start and the end of the write (i.e. a split finished in the meantime)

Reply:
Done.